Sensor Class

fuzzybear3 opened the pull request:
Sensor Class incuding .h and .cpp file. Created LFRobot namespace.

Commit:
Initial Commit

Commit:
Sensor Class V2

Added default pinMode in Sensor constructor.

blhough commented on Sensor.cpp:
```diff
@@ -0,0 +1,31 @@
+		if (digitalRead(pin) == LOW)
+		{
+			return true;
+		}
```
if pin is HIGH then no value is returned

blhough replied:
`return digitalRead(pin) == low;`
handles both cases in one line

fuzzybear3 replied:
clever...

blhough commented on Sensor.cpp:
```diff
@@ -0,0 +1,31 @@
+{
+	Sensor::Sensor(int inPin)
+	{
+		pinMode(pin, INPUT);
```
careful, you use `pin` before you initialize it.
It's best to use an initializer list to guarantee the variable is initialized before you use it

fuzzybear3 replied:
what is an initializer list?

blhough replied:
member initializer list:
http://www.learncpp.com/cpp-tutorial/8-5a-constructor-member-initializer-lists/

Commit:
sensor class V2.01

small changes.

Commit:
Sensor class V2.02

Commit:
Added documentation + Initialization List

Added basic documentation describing the use of an initialization list
in the Sensor constructor function and the return of  the bool
comparisons in isLow().

Commit:
Sensor ClassV2.04

Commit:
starting SensorArray class

Commit:
forgot to save files before commit...

blhough commented:
Lets keep this PR to just the sensor class and leave the sensor array for another one

blhough commented on Sensor.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <Sensor.h>
```
`#include "Sensor.h"`

Commit:
random save

Commit:
sensorArray class testing V .01

not working.

Commit:
updating gitignore.

Commit:
saving progress(not working)

Commit:
SensorArray Class V 0.2

Testing

Commit:
Merge remote-tracking branch 'refs/remotes/origin/dev' into kyle-sensor-class

Commit:
Merge branch 'kyle-sensor-class' of https://github.com/SRJC-Computer-Science-Club/line-follower into kyle-sensor-class

blhough commented on Sensor.cpp:
```diff
@@ -0,0 +1,29 @@
+			pinMode(pin, OUTPUT);
+			digitalWrite(pin, HIGH);
+		}
+		if (mode == IN)
```
Though the compiler will surely optimize this, `mode` can never equal `OUT` and `IN` so an `else if` or just an `else` (since mode only has the two values) would be better here.

blhough commented on Sensor.h:
```diff
@@ -0,0 +1,21 @@
+
+		private:
+
+			int pin;
```
should be constant to prevent any accidental changing

blhough commented on SensorArray.h:
```diff
@@ -0,0 +1,29 @@
+	
+
+	private:
+		Sensor* sensorArray[8];
```
use `NUMBER_OF_SENSORS`

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+
+		for (int i = 0; i < NUMBER_OF_SENSORS; i++)
+		{
+			sensorArray[i] = new Sensor(pins[i]);
```
regarding `pins[]`, this is really dangerous because there is nothing guaranteeing that `pins` is at least size `NUMBER_OF_SENSORS`. If someone calling this function only supplies a `pins` of size 7 then this would read out of bounds and you wouldn't know it (until you called Sensor::setMode(OUT) and potentially set a random pin high and fry something).

Ideally you would make a new class that enforces that the right number of pins is provided (or use std::vector).

blhough commented on Sensor.cpp:
```diff
@@ -0,0 +1,29 @@
+		if (mode == OUT)
+		{
+			pinMode(pin, OUTPUT);
+			digitalWrite(pin, HIGH);
```
If `Sensor` is only going to be used with the Reflectance Sensor Array then this is fine. However if `Sensor` is to be reused by other sensors types in the project, defaulting the output to HIGH is risky.

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+{
+	SensorArray::SensorArray(int pins[])
+	{
+		Sensor* sensorArray[NUMBER_OF_SENSORS];
```
This is a re-declaration of the member variable at SensorArray.h:23. C++ allows this by hiding (shadowing) the variable with the wider scope ([read more about it here](https://en.wikipedia.org/wiki/Variable_shadowing)). So what this means is that when you assign to `sensorArray` on line 5 you are assigning to the local one, not the class level one. So basically a major memory leak when `sensorArray` goes out of scope at the end of the constructor. Every other use of `sensorArray` in the class is using the un-initialized one (bad).


On another note, I know you just learned about pointers, so is good time to learn when pointers are not needed. Notice in your implementation you never move a pointer around, reassign it, pass it to a function etc. You aren't using many of the benefits of pointers but are still stuck with all the negatives (null dereference, memory leak, etc.). The one benefit you are using is that you can delay creating the objects in the array until you are able to put meaningful objects in it, but this is [generally](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c43-ensure-that-a-class-has-a-default-constructor) not considered a good enough reason to use pointers. It is best to just use an array of objects (no pointers) and then provide a default constructor for `Sensor`

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+	{
+		for (int i = 0; i < NUMBER_OF_SENSORS; i++)
+		{
+			delete sensorArray[i];
```
If `sensorArray` were to be properly initialized (see above) then this is the correct way to handle the de-allocations 🥇

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+
+		int SensorArray::getLineOffset()
+		{
+			27;
```
This probably works by some weird C backwards compatibility idk, but it's not good form.  Just make it a const int LINE_OFFSET if that is all you need

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+			sensorArray[i]->setMode(OUT);
+
+		}
+	}
```
Don't forget to allow for 10us after setting high. It is possible that the code already takes that long to go from prepSensor to readSensorValues

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+				if (tempSensorValues[sensorNum] == 0)// if array element is empty.
+				{
+
+					if (sensorArray[sensorNum]->isLow) // if low, record time.
```
`isLow` is a function, need ()

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+					{
+
+						endTime = micros();
+						lengthOfTime = endTime - startTime;
```
Are these really necessary ?
could make it
`tempSensorValues[sensorNum] = micros() - startTime;`

blhough commented on SensorArray.cpp:
```diff
@@ -0,0 +1,94 @@
+
+
+	void SensorArray::readSensorValues()
+	{
```
If this ends up being too slow/inacurrate I thought of a really fast way using bit operations

Commit:
V0.02 (bool convert error)

Commit:
Loop not looping.

read comments in .ino file to see where the loop stops looping.

Commit:
weird  fail

Commit:
Merge remote-tracking branch 'origin/kyle-sensor-class' into kyle-sensor-class

# Conflicts:
#	Sensor.cpp
#	SensorArray.cpp
#	line-follower.ino

theDataSmith committed:
Fixed project files

Commit:
its working, maybe.
